src/consumer.py

```diff
@@ -0,0 +1,89 @@
+import asyncio
+import os
+import struct
+from datetime import datetime
+
+import typer
+from fastapi import FastAPI
+from prometheus_client import Counter, Histogram, make_asgi_app
+from streamstore import S2
+from streamstore.schemas import SequencedRecord
+from uvicorn import Config, Server
+
+S2_AUTH_TOKEN = os.getenv("S2_AUTH_TOKEN")
+
+METRICS_NAMESPACE = "stress"
+
+e2e_latency_histogram = Histogram(
+    name="e2e_latency_seconds",
+    documentation="End-to-end latency in seconds. This measures the time taken from when a record is produced "
+    "(by the producer) to when it is read by the consumer.",
+    namespace=METRICS_NAMESPACE,
+)
+read_records_counter = Counter(
+    name="read_records_total",
+    documentation="Total number of records successfully read.",
+    namespace=METRICS_NAMESPACE,
+)
+read_bytes_counter = Counter(
+    name="read_bytes_total",
+    documentation="Total number of bytes successfully read.",
+    namespace=METRICS_NAMESPACE,
+)
+read_errors_counter = Counter(
+    name="read_errors_total",
+    documentation="Total number of failed read session operations.",
+    namespace=METRICS_NAMESPACE,
+)
+
+
+def observe_metrics(record: SequencedRecord):
+    name, val = record.headers[0]
+    read_bytes_counter.inc(len(record.body) + len(name) + len(val))
+    created_ts = struct.unpack("d", val)[0]
+    e2e_latency = (datetime.now() - datetime.fromtimestamp(created_ts)).total_seconds()
+    e2e_latency_histogram.observe(e2e_latency)
+
+
+async def consumer(basin_name: str, stream_name: str):
+    async with S2(auth_token=S2_AUTH_TOKEN) as s2:
+        stream = s2[basin_name][stream_name]
+        start_seq_num = await stream.check_tail()
+        with read_errors_counter.count_exceptions():
```

Comment:
Hmm does this retry in case of an exception..? There won't be a chance for Prometheus to scrape an error if we ash and there is a higher-level container restart. So I think should retry internally.

Reply:
no it doesn't. good catch. wanted to revisit this one on this angle.

Reply:
din go with a retry. just ensuring prometheus gets a chance to see that error occurred through a sleep. i think that should be okay. lemme know if you think otherwise.